src/Test/Plutip/Contract.hs

```diff
@@ -1,54 +1,95 @@
+  to run assertions against result of contract execution and funds at the wallet's UTxOs after contract being run.
+  
+  Each test case starts with `assertExecution`, which accepts:
+    - description of test case
+    - initial funds distribution at wallets addresses (with optional Value assertions to be performed after the Contract run)
+    - contract to be tested (passed to `withContract`, more on this later)
+    - list of assertions to run against Contract return result, observable state and/or error
+  
   At least one TestWallet is required, this will be used as the own wallet for the contract. Any other
   wallets can be used as other parties in transactions.
 
   A TestWallet can be initialised with any positive number of lovelace, using the `initAda` or
   `initLovelace`. In addition, the value in these wallets can be asserted after the contract
-  execution with `initAdaAssertValue` or `initAndAssertAda`.
+  execution with `initAdaAssertValue` or `initAndAssertAda`. When `initAdaAssertValue` or `initAndAssertAda` used
+  to initiate wallets corresponding test case will be added automatically.
+
+  Each assertion in assertions list will become separate test case in `TestTree`,
+  however Contract will be executed only once.
+
+  E.g.:
 
- > shouldSucceed "Get utxos" (initAda 100) $ withContract $ \_ -> do
- >   pkh <- Contract.ownPaymentPubKeyHash
- >   utxosAt $ pubKeyHashAddress pkh Nothing
+ > assertExecution 
+ >   "Some Contract"                   -- Contract description
+ >   (initAda 100)                     -- wallets and initial funds for them (single wallet in this case)
+ >   (withContract $ \_ -> myContract) -- contract execution
+ >   [ shouldSucceed                   -- list of assertions
+ >   , not $ shouldYield someResult
+ >   , stateSatisfies "description" somePredicate
+ >   ]
 
   To use multiple wallets, you can use the `Semigroup` instance of `TestWallets`. To reference the
-  wallet inside the contract, the following callback function is used, when
-  supplying a contract toa test case: @[PaymentPubKeyHash] -> Contract w s e a@.
-  Note that @[PaymentPubKeyHash]@ does not include the contract's own wallet, for that you can use `Plutus.Contract.ownPaymentPubKeyHash` inside the Contract monad.
-
- > shouldSucceed "Send some Ada" (initAda 100 <> initAndAssertAda 100 110) $ withContract $
- >   \[pkh1] -> submitTx (Constraints.mustPayToPubKey pkh1 (Ada.lovelaceValueOf amt))
-
-If you have multiple contracts depending on each other, you can chain them together using
-
-`withContract` and `withContractAs`.
- > shouldSucceed
- >   "Two contracts after each other"
- >   (initAndAssertAdaWith 100 VLt 100 <> initAndAssertAdaWith 100 VLt 100)
- >   $ do
- >     void $
- >       withContract $
- >         \[pkh1] -> payTo pkh1 10_000_000
- >     withContractAs 1 $
- >       \[pkh1] -> payTo pkh1 10_000_000
+  wallet inside the contract, the following callback function is used together with `withContract`:
```

Comment:
Just checked with `cabal haddock`, but the newline doesn't seem to be rendered there. If that was your intention, could you take a look?

Reply:
I've revisited and updated documentation and made fixes. Haddock generated docs should be OK now.